src/shared/edit-util.c

```diff
@@ -0,0 +1,378 @@
+                        } else
+                                free_and_replace(args[0], e);
+
+                        execvp(name, (char* const*) args);
+
+                        /* We do not fail if the editor doesn't exist because we want to try each one of them
+                         * before failing. */
+                        if (errno != ENOENT) {
+                                log_error_errno(errno, "Failed to execute '%s': %m", name);
+                                _exit(EXIT_FAILURE);
+                        }
+                }
+
+                log_error("Cannot edit files, no editor available. Please set either $SYSTEMD_EDITOR, $EDITOR or $VISUAL.");
+                _exit(EXIT_FAILURE);
+        }
+
+        return 0;
+
+oom:
+        log_oom();
+        _exit(EXIT_FAILURE);
+}
+
+int trim_edit_markers(const char *path, const char *marker_start, const char *marker_end) {
+        _cleanup_free_ char *old_contents = NULL, *new_contents = NULL;
+        char *start, *end, *stripped;
+        int r;
+
+        /* Trim out the lines between the two markers */
+        r = read_full_file(path, &old_contents, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read temporary file \"%s\": %m", path);
+
+        start = strstr(old_contents, marker_start);
+        if (start)
+                start += strlen(marker_start);
+        else
+                start = old_contents;
+
+        end = strstr(start, marker_end);
+        if (end)
+                *end = '\0';
+
+        stripped = strstrip(start);
+        if (isempty(stripped))
+                return 0; /* All gone now */
+
+        new_contents = strjoin(stripped, "\n"); /* Trim prefix and suffix, but ensure suffixed by single newline */
+        if (!new_contents)
+                return log_oom();
+
+        if (streq(old_contents, new_contents)) /* Don't touch the file if the above didn't change a thing */
+                return 1; /* Unchanged, but good */
+
+        r = write_string_file(path, new_contents, WRITE_STRING_FILE_CREATE|WRITE_STRING_FILE_TRUNCATE|WRITE_STRING_FILE_AVOID_NEWLINE);
+        if (r < 0)
+                return log_error_errno(r, "Failed to modify temporary file \"%s\": %m", path);
+
+        return 1; /* Changed, but good */
+}
+
+int edit_files_add(
+                EditFile **ef,
+                const char *path,
+                const char *original_path,
+                char * const *comment_paths) {
+
+        EditFile edit_file = {};
+        size_t n;
+
+        assert(ef);
+        assert(path);
+
+        n = edit_files_count(*ef);
+
+        if (!GREEDY_REALLOC0(*ef, n + 2))
+                return log_oom();
+
+        edit_file.path = strdup(path);
+        if (!edit_file.path)
+                return log_oom();
+
+        if (original_path) {
+                edit_file.original_path = strdup(original_path);
+                if (!edit_file.original_path)
+                        return log_oom();
```

Comment:
When this is triggered, `edit_file.path` is leaked. I do not think it is necessary to introduce a temporal variable `edit_file`, but simply assign `(*ef)[n].path` or so.